### DISCUSSION TITLE
Help with timeouts

### DISCUSSION THREAD

**Discussion by JamesLampert** [reach-sh/reach-lang#814] [Q&A] [answered]
After doing the Rock/Paper Scissors tutorial (using Ms. Hodge's "Ten Days of Blockchain" video series),
I've been trying to work through the Reach Workshops, trying to learn as much as I can from them.

I'm stuck in the first one, "Hash Lock." And I'm in a local devnet.

I'm trying to add a timeout, so that if "Bob" doesn't retrieve the monopoly Algos within a certain time limit, they revert back to "Alice."

Now obviously in a real world scenario, in which "Bob" and "Alice" aren't sharing a single instance of a common frontend, "Alice" would be able to do this manually. But that has no bearing on the observed behavior.

The first thing I added, once I had achieved basic functionality with Hash Lock was to give "Bob" an opportunity to try again:

```
    var [pwdGood, count] = [false, 3];
    invariant (balance() == amount && count >= 0)
    while (pwdGood == false && count > 0) {
        commit();
        Bob.only(() => {
          const pwd = declassify(interact.getPassword());
          const bobDigest = digest(pwd)
        })
        Bob.publish(bobDigest)
        if (passDigest != bobDigest && count > 1) Bob.interact.badPassword(count-1);
        [pwdGood, count] = [(passDigest == bobDigest), count-1];
        continue
    }
    Bob.interact.finalPassword();
    if(pwdGood) {
        transfer(amount).to(Bob);
    }
    else transfer(balance()).to(Alice)
    commit();
```

It took me a lot of experimentation, and a lot of guidance from the Reach Help channel of Discord, but I got this working, by adding an empty "Bob.publish();" just before this.

So then, I added a timeout to the above, i.e.,
```
        Bob.publish(bobDigest).timeout(relativeTime(2), () => closeTo(Alice, Alice.interact.informTimeout));
```
and added an informTimeout to the frontend.

So now, my frontend looks like this:
```
await Promise.all([
    backend.Alice(ctcAlice, {
        amount: theAmount,
        password: thePass,
        informTimeout: () => {console.log('There was a timeout.');}
    }),
    backend.Bob(ctcBob, {
        getPassword: async() => {
         //   for ( let i = 0; i < 10; i++ ) {
         //       console.log(`  Bob takes his sweet time...`);
         //       await stdlib.wait(1);
         //     }
            bobPass = await ask('Password, Bob?', parseInt);
            return bobPass;
        },
        badPassword: (count) => {
            if (count > 1) console.log(`Wrong password. You may try ${count} more times.`);
            else console.log('Wrong password. You may try ONCE more.');
        },
        finalPassword: () => {
            console.log(`Alice gave: ${thePass} Bob gave: ${bobPass}.`)
        }
    }),
]);
```

If I uncomment the block of four commented-out lines, so that the frontend itself delays things, the timeout fires (and fires very quickly). And it fires whether I put that block before or after "bobPass = await ask('Password, Bob?', parseInt);"

But if those lines are commented out, and I simply walk away while it's in the ask, then it just sits there, and never times out.

Since the timeout-related stuff is borrowed from the Rock Paper Scissors tutorial (which I'd modified to run in TestNet, with existing accounts), I tried to bring that up again, to compare its behavior, but it has stopped working (and I'll put the details of that in a separate thread).

**Comment by jeapostrophe** [comment, marked as answer]
A few comments...

First, even without your change, Bob has an infinite number of tries to give the password, because he can check if something is the correct password entirely locally without ever interacting with the contract. So you haven't added any functionality with your new... you've just done more work and made the program more complicated. I recommend that you remove that if you are going to really do something with this.

Second, the reason you never actually get a timeout is that when you run a devnet, literally nothing happens until you prod the network by doing something. Thus, time really has not progressed so it is still "on time". That's why in a test scenario (like with a devnet), you use `stdlib.wait` to simulate the natural passage of time on the real network. If you were to have run this program on a real network, there'd basically never be any way that you'd succeed, because you probably don't have a fast enough connection to a node to get something within two blocks.

**Reply by JamesLampert** [reply]
Well, the only thing I'm actually doing with this is trying to learn as much as I can from it before moving on to "Relay Account." And if it led me to the realization that time doesn't actually pass on a devnet without _something_ "prodding" it (I suspected this, which is why I dug out Rock Paper Scissors), and it also led me to realize how much I _don't_ understand about control flow in Rock Paper Scissors, then it's served its purpose.

And I shortened the timeout here (I started it out much longer) to something insanely short, _precisely_ because I was trying to get it to fail.
